### ISSUE TITLE
Allow using vim-native gq instead of LSP provided one

### ISSUE THREAD

**Issue by reegnz** [neovim/neovim#22611]
### Problem

In #19677 backward compatibility was broken, as previously `gq` was vim-native, even with an LSP configured. After that PR though, `gq` is now set by the LSP. This should have been opt-in to maintain backward compatibility, instead of opt-out with some `on_attach` hackery that isn't even documented in the neovim docs.
There does not seem to be an option to opt-out of setting `gq` to use the LSP provided mechanism.
The documentation to `gq` also fails to mention this LSP detail, so it was only after searching github issues I've found out about this nuanced detail.

### Steps to reproduce

* Install the https://github.com/bash-lsp/bash-language-server
* open a bash file
* try doing `gq` on a block comment
* be disappointed that `gq` does not work anymore. :(

### Expected behavior

Overriding `gq` from the LSP should be opt-in when you configure the LSP.
Alternatively, one should be able to opt-out with a mechanism that is documented in the neovim docs about  with a neovim setting to not use this feature.
Currently only the `lsp` part of the docs mention it, and because of that a user looking at the `gq` or `formatexpr` docs can only scratch their heads about what has changed, since they only upgraded neovim, and what used to work does not work anymore.

### Neovim version (nvim -v)

0.8.3

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 13.2.1

### Terminal name/version

iTerm 3.4.19

### $TERM environment variable

xterm-256color

### Installation

brew install neovim

**Comment by reegnz** [comment]
My biggest gripe really is that the docs failed to help me in finding what broke. So I guess if the `formatexpr` docs would refer to the `lsp` behaviour that might override it's default, that would definitely help people.

I'm also seeing that people come up with ways to opt out of this, for example here: https://github.com/jose-elias-alvarez/null-ls.nvim/issues/1131

So it might make sense documenting how you can simply opt out of this lsp feature on a case-by-case basis.
The proposed solutions seem to be quite complicated, so maybe some development is also required besides doc improvements, that is why I mainly opened the ticket.

**Comment by reegnz** [comment]
After digging a bit, it seems like LSP-s cannot even report whether they can do  rangeformat or not on a file-by-file basis, so LSP-s supporting multiple filetypes, like efm-langserver or null-ls have to report formatting capabilities even if they do formatting for just one of their supported file-types.

I think therefore making `formatexpr` default to the lsp formatter was a bit premature.
The LSP spec itself cannot guarantee that the LSP is capable of formatting all supported file types the LSP works with, as the LSP has to report formatting capability even if it only supports formatting for a subset of filetypes it works with.

See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize for details.

Essentially some LSP-s advertize capabilities as a "might be able to do it for the file, but there's no harm if it doesn't". The referenced PR in the issue however makes the assumption that LSP-s that report formatting capabilities can format for all file types they can be registered for, which is simply not the case.

This leads to the weird quirks like `gq` working with null-ls if you have black defined, but if you only have a linter defined for bash, then the neovim default breaks the basic formatting capability of `gq` leaving the user with nothing.

I have to emphasize that I think this is an LSP spec issue, but because it works like that today, defaulting `gq` based on LSP formatting capability is a mistake.

**Comment by mfussenegger** [comment]
- LSP is already opt-in, if you opt-in, you opt in to it's functionality, that includes formatting. There were lots of complaints about having to setup dozens of keymaps. We have to choose one option and can't make everyone happy
- There is `gw` as alternative to `gq` which doesn't rely on the `formatexpr`
- The bash-language-server doesn't advertise formatting capabilities and `formatexpr` is empty after starting a client for the bash-language-server
- That LSPs may not have capabilities for _all_ filetypes is more of an issue specific to null-ls (or other "fake" language-servers). It's hardly the case with regular language servers.

You can opt-out by resetting `formatexpr` in a `LspAttach` event. The `:help lsp-quickstart` doc mentions the behavior and it's assumed that anyone ussing LSP would read at least the LSP intro.

A PR to extend the section about the defaults with how it can be reset is welcome.

**Comment by justinmk** [comment]
As mentioned, does `gw` not do what you want? 

Besides @mfussenegger 's suggestions, agree that we can add hints to `:help gq` and similar.

**Comment by reegnz** [comment]
Yes, probably going to transition over to `gw`. But the lack of docs about this behaviour, so thankfully I have an escape-route.

**Comment by reegnz** [comment]
Regarding bash lsp -> I thought it was that, but at the end I realized the behaviour is caused by using either `efm-langserver` or `null_ls` as explained in my later comment.

**Comment by reegnz** [comment]
I opened a PR as an initial improvement to the docs.

**Comment by ibash** [comment]
Just noting this was my biggest gripe with neovim lsp -- hard to overwrite 10+ years of muscle memory.
